cmd/syncComparator.go

```diff
@@ -54,30 +199,191 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
-	if present {
-		defer delete(f.sourceIndex.indexMap, destinationObject.relativePath)
-		if f.disableComparison || sourceObjectInMap.isMoreRecentThan(destinationObject) {
-			err := f.copyTransferScheduler(sourceObjectInMap)
-			if err != nil {
-				return err
+	lcFolderName = filepath.Dir(lcRelativePath)
+	lcFileName = filepath.Base(lcRelativePath)
+
+	f.sourceFolderIndex.lock.Lock()
+
+	foldermap, folderPresent := f.sourceFolderIndex.folderMap[lcFolderName]
+
+	// Do the auditing of map.
+	defer func() {
+		if folderPresent {
+			if len(f.sourceFolderIndex.folderMap[lcFolderName].indexMap) == 0 {
+				size := int64(unsafe.Sizeof(objectIndexer{}))
+				atomic.AddInt64(&f.sourceFolderIndex.totalSize, -size)
+				delete(f.sourceFolderIndex.folderMap, lcFolderName)
 			}
 		}
-	} else {
-		// purposefully ignore the error from destinationCleaner
-		// it's a tolerable error, since it just means some extra destination object might hang around a bit longer
-		_ = f.destinationCleaner(destinationObject)
+		f.sourceFolderIndex.lock.Unlock()
+	}()
+
+	// Folder Case.
+	if destinationObject.isVirtualFolder || destinationObject.entityType == common.EEntityType.Folder() {
+		if destinationObject.isFolderEndMarker {
+			// End Marker marks enumeration complete of the folder.
+			// End Marker come in following cases:-
+			// 1. Dir not present on target side, then we need to create directory and files.
+			// 2. Dir present and some files are added.
+			// 3. Dir present and some files are modified.
+			// For each case we need to process left over files in ObjectIndexerMap.
+			if folderPresent {
+				if lcRelativePath != "" {
+					// We should not find this entry as it should be removed as part of parent enumeration only exception to this is root folder (that's what we checking in if statement).
+					// But due to go routine scheduling it may happen before parent enumeration at target finish or even start, other goroutine picked the child entry and finish the enumeration.
+					// In that case we don't need to do anything. If folder is empty and don't exist then will be created as parent enumeration done.
+					// If folder is not empty then it will be created by it's children. Only thing, it's metadata will be set by parent only as it has the required info.
+					fmt.Printf("lcRelativePath[%s], We should not find this entry as it should be removed as part of parent enumeration."+
+						"\n But due to go routine scheduling it may happen before parent enumeration at target finish child done with enumeration."+
+						"\n If folder is empty and don't exist then will be created as parent enumeration done. Otherwise will be created by its child enumeration."+
+						"\n Its metadata will be set by parent only as it has the required info.", lcRelativePath)
+				}
+			}
+
+			lcFolderName = path.Join(lcFolderName, lcFileName)
+			foldermap, folderPresent = f.sourceFolderIndex.folderMap[lcFolderName]
+			if folderPresent {
```

Comment:
No Tomar , ["dir1/dir2"]["."] will be removed by getStoredObject, so if all file present on source and target same, folder will be deleted. Like in case if someone create the file and delete it  before next sync. Ctime changed and we do target traverse.

Reply:
As discussed let's delete folderMap[folder] only from FinalizeTargetDirectory since we finalize a directory when we are done processing all its children, so that's the right time to delete it from folderMap.